Module Health: Node Manager: First Iteration

tommyp1ckles opened the pull request:
This PR seeks to provide modular health status checking on the "node manager" component of the Agent. Specifically the goals are:

## Intentions of this PR:

### Bubble-up all errors in Linux node handler code

Node Manager related code relies heavily on logging (or just ignoring) errors far down the stack, instead of bubbling them up.

The result of this is:
* Logged errors have little context on the root event causing them to be invoked.
* It's hard to trace failure from a particular event relating to all reconcile procedures that need to run.
* It's hard to definitively to say if reconciliation for a Node Event was successfully.

The changes in this PR seek to address this by handling and bubbling up all relevant errors that may be encountered while doing Linux Node update reconciliation logic.

As well, this PR seeks to **avoid** any changes to the control-flow logic of the reconciliation by _not_ immediately returning at places where errors may occur.
Rather, this follows a pattern of combining lists of errors into relevant/scoped multi errors (using the multierr library). All code will execute as before, with the difference being that errors are handled up the stack.

### Use Hive's Modular Health Reporter to report status of node manager.

This is a first attempt at trying to manage health status for a module. 

#### Follow up work:
* Investigate Reconciliation: Break Linux node logic into atomic and idempotent tasks that can be executed and retried (+ reported on more easily).

tommyp1ckles committed:
Add Name() function to NodeHandler interface to identify handlers.

The NodeHandler interface is used to chain handlers to be executed when Node updates occur.
We want to be able to better identify these handlers when they fail so we
collect and report on errors we can provide more context on where the
error occurred.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
node/manager: add health reporter and track errors from node manager

Inject cell.HealthReporter into module and use it to do a basic health
check on the modules periodic background sync.

The background sync performs a "validation" call that essentially tries
to re-reconcile all necessary changes for existing nodes in the manager.
This is performed every minute.

For now, we will ignore the Update/Create event handlers on the manager,
as the background sync should give a good indication of the state
without alternating between healthy/ok too fast.

If the background validation succeeds on all nodes, we consider the node
manager "healthy", otherwise we report a degraded state. This simplified
model is nice because we only have to consider one source of errors,
the validation loop.

One thing this change omits is health checking on NodeDelete failures,
node deleted is a bit more difficult since failed node deletes are not
retried - possibly leaving leftover configuration.
This would mean that an error in the delete would leave the node in a
permanently degraded state, as well tracking errors here may result in
unbounded memory growth as nodes deleted in a cluster with high churn
may fail and never be resolved.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
datapath/linux/node: nodeUpdate aggregate errors for enable IPSec.

enableIPSec run when invoking updateNode(...), which is called from the
Update/Add/Validate node handler interface implementations for the
datapath linux node handler (linuxNodeHandler).

This refactors enableIPSec and all underlying functions to return an accumulated error upon failures.
IPSec enable exection will continue even if error is collected as to not affect control flow.

Top level enableIPSec error is returned wrapped to scope what part of the updateNode procedure failed.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
datapath/linux/node: aggregate errors for node route update.

updateOrRemoveNodeRoutes() attempts to reconcile a diff between old and new routes, depending on routing mode.
Currently procedure tasks to apply the ip route change via netlink can fail and will not have their error logged or returned.

This updates the updateOrRemoveNodeRoutes code paths such that all errors are bubbled up and returned without affecting execution flow. Existing logging is preserved.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
datapath/linux/node: add comment explaining neighbor error handling.

Neighbor handling is a bit tricky as it's done async, possibly in
parallel with other neighbor adds.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
datapath/linux/node: fix fmt.Errorf linting issues.

Use wrap directives instead of string for correct error
handling and fix capitalized error messages.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
hive/hivetest: add mock health reporter.

A mock implementation of cell.HealthReporter will be useful in the
future for testing as more modules depend on being passed a reporter.

This implementation simply keeps the last msg/err/stopped state.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
linux/ipsec/ipsec_linux: return errors from ipsec/xfrm functions.

Previously many ip xfrm netlink requests would fail, and either be logged or
have the error unhandled.

This adds error handling in xfrm commands that are used by procedures in linux datapath updateNode.

Eventually, we'll want to correctly bubble up all meaningful errors in this package.
This only changes functionality that affects linux node datapath update/delete/create logic.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
datapath/linux/node_ids: return errors from node ID allocate funcs.

The Node ID allocation functions in node_id.go don't currently return
errors when they fail, instead just logging the error.

This changes these to return an error, while not changing the behavior
of the allocate code.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
datapath/linux/node: prevent errs being logged for benign route err.

Top level node manager logging added in previous commits was failing CI
as we are now logging errors that would previously be swallowed.

In this case, the problem comes from the attempt to install routes
always being attempted and possibly failing on one of the ip families
due to a nil net.IP being returned for the local IP.

In the interest of not making changes to control logic, we will handle
this case separately to prevent such errors from bubbling up.

As a follow up, these calls should be refactored such that only route
update requests that are expected to succeed are attempted. Which will
allow us to return/log more meaningful errors.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

chancez reviewed (approved):
Hubble side looks good.

tommyp1ckles committed:
datapath/linux/node: aggregate errors for subnet ipsec.

Aggregate and return all errors encountered while performing ipsec: route procedure.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
datapath/linux: aggregate NodeDeleted event handler errors.

Similar to node update handler, node deleted needs to tear down some
host configuration state during a node delete.

This aggregates all errors returned in the NodeDelete procedure, and
returns error to node manager.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

pippolo84 commented on pkg/datapath/linux/node.go:
```diff
@@ -500,7 +500,7 @@ func (n *linuxNodeHandler) NodeUpdate(oldNode, newNode nodeTypes.Node) error {
 }
 
-func upsertIPsecLog(err error, spec string, loc, rem *net.IPNet, spi uint8) {
+func upsertIPsecLog(err error, spec string, loc, rem *net.IPNet, spi uint8, accs ...*error) {
```
I think it'd be better to add a new separate helper to wrap errors with additional scope, instead of piggyback that on `upsertIPsecLog`. That might also be a chance to add a comment and explain why we are doing that.

tommyp1ckles replied:
I wanted to piggyback on this just to reduce the number of changes for now, I agree with you and @ti-mo that the accs pointer approach isn't very idiomatic so ill change this to behave more similarly to errors.Join.  This should also make it more clear in line that the error is being pushed on to the errs list.

The error context being added isn't always that great atm, so I think we should defer this for now and add more helpful context more thoughtfully in the future.

ti-mo commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -517,42 +518,48 @@ func ipSecReplacePolicyOut(src, dst *net.IPNet, tmplSrc, tmplDst net.IP, nodeID
 	if err != nil {
 		scopedLog.WithError(err).Warning("Failed to list XFRM states for deletion")
-		return
+		return acc
```
This will return `nil` if `netlink.XfrmStateList` returns error. Looks like `ipsecDeleteXfrmPolicy` below gets it (mostly) right.

ti-mo commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -517,42 +518,48 @@ func ipSecReplacePolicyOut(src, dst *net.IPNet, tmplSrc, tmplDst net.IP, nodeID
 			}
 		}
 	}
+	return err
```
```suggestion
	return acc
```
?

pippolo84 commented on pkg/datapath/linux/node.go:
```diff
@@ -544,7 +552,8 @@ func (n *linuxNodeHandler) enableSubnetIPsec(v4CIDR, v6CIDR []*net.IPNet) {
 // encryption = disabled).
-func (n *linuxNodeHandler) encryptNode(newNode *nodeTypes.Node) {
+func (n *linuxNodeHandler) encryptNode(newNode *nodeTypes.Node) error {
+	var acc error
```
Nit: I see that `acc` is short for "accumulator" but I still find it a bit odd. Maybe `errs` might sound more idiomatic and familiar for the casual reader?

tommyp1ckles replied:
yeah, I think you and @ti-mo are right 😄 , it seemed like a good choice at the time but I think errs is what most people would expect and thus will be more readable

ti-mo commented on pkg/hive/hivetest/health.go:
```diff
@@ -0,0 +1,52 @@
+	return m.err
+}
+
+func (m *MockHealthReporter) IsStopped() bool {
```
This is personal preference, but IMO this reads a little more naturally and makes for a shorter method name:
```
if m.Stopped() {
  do something
}
```

```suggestion
func (m *MockHealthReporter) Stopped() bool {
```

ti-mo commented on pkg/hive/hivetest/health.go:
```diff
@@ -0,0 +1,52 @@
+	m.err = nil
+}
+
+func (m *MockHealthReporter) Degraded(msg string, err error) {
```
Like above:

```suggestion
func (m *MockHealthReporter) Degrade(msg string, err error) {
```

tommyp1ckles replied:
See: comment below

ti-mo commented on pkg/hive/hivetest/health.go:
```diff
@@ -0,0 +1,52 @@
+	m.err = err
+}
+
+func (m *MockHealthReporter) Stopped(msg string) {
```
```suggestion
func (m *MockHealthReporter) Stop(msg string) {
```

tommyp1ckles replied:
Re: this suggestion and the two above I think are more pointed at the hive/cell/HealthReporter interface.

I'm going to re-frame from changing this as part of this PR, I don't think that API has stabilized yet so we can still make changes.  I'm working on some design proposals that will likely include this API so I think its best we leave that to there.

ti-mo commented on pkg/node/manager/manager.go:
```diff
@@ -133,7 +141,9 @@ func (m *manager) Subscribe(nh datapath.NodeHandler) {
 		v.mutex.Lock()
-		nh.NodeAdd(v.node)
+		if err := nh.NodeAdd(v.node); err != nil {
+			log.WithError(err).Errorf("initial subscribe initial node handler %s failed on node %s", nh.Name(), v.node.Name)
```
This stutters a bit, `initial subscribe to node handler`?

ti-mo commented on pkg/node/manager/manager.go:
```diff
@@ -272,10 +283,15 @@ func (m *manager) backgroundSyncInterval() time.Duration {
 		syncInterval := m.backgroundSyncInterval()
 		log.WithField("syncInterval", syncInterval.String()).Debug("Performing regular background work")
 
+		var errs error
```
Why use `errs` here but `acc` everywhere else? We should converge on a single convention. Personally, I prefer `errs`.

ti-mo commented on pkg/node/manager/manager_test.go:
```diff
@@ -829,3 +840,51 @@ func (s *managerTestSuite) TestNode(c *check.C) {
+
+	// Start the manager
+	assert.NoError(m.Start(context.Background()))
+	time.Sleep(time.Millisecond * 100)
```
time.Sleep() is unreliable, can this be done without? Consider adding a readiness notification to the manager to both speed up the test and make it less flaky.

tommyp1ckles replied:
agreed, think this is an artifact at my first attempt where I used sleeping instead of syncing up with the sync loop.

ti-mo commented on pkg/datapath/linux/node.go:
```diff
@@ -176,9 +190,10 @@ func cidrNodeMappingUpdateRequired(oldCIDR, newCIDR cmtypes.PrefixCluster, oldIP
+	var acc error
 	if !oldCIDR.IsValid() {
-		return
+		return acc
```
I'd explicitly `return nil` here and declare `acc` closer to where it's used.

ti-mo commented on pkg/datapath/linux/node.go:
```diff
@@ -176,9 +190,10 @@ func cidrNodeMappingUpdateRequired(oldCIDR, newCIDR cmtypes.PrefixCluster, oldIP
 
-func deleteTunnelMapping(oldCIDR cmtypes.PrefixCluster, quietMode bool) {
+func deleteTunnelMapping(oldCIDR cmtypes.PrefixCluster, quietMode bool) error {
+	var acc error
```
I don't think there can be multiple errors in this function.

ti-mo commented on pkg/datapath/linux/node.go:
```diff
@@ -272,14 +289,16 @@ func installDirectRoute(CIDR *cidr.CIDR, nodeIP net.IP) (routeSpec *netlink.Rout
 		// in quiet mode as we don't know whether it exists or not
 		if firstAddition {
-			n.deleteAllDirectRoutes(newCIDRs, newIP)
+			acc = errors.Join(acc, n.deleteAllDirectRoutes(newCIDRs, newIP))
```
I'd probably do `return n.deleteAll...` here.

ti-mo commented on pkg/datapath/linux/node.go:
```diff
@@ -272,14 +289,16 @@ func installDirectRoute(CIDR *cidr.CIDR, nodeIP net.IP) (routeSpec *netlink.Rout
+			acc = errors.Join(acc, n.deleteAllDirectRoutes(newCIDRs, newIP))
 		}
-		return nil
+		return acc
```
.. and `return nil` here.

ti-mo commented on pkg/datapath/linux/node.go:
```diff
@@ -272,14 +289,16 @@ func installDirectRoute(CIDR *cidr.CIDR, nodeIP net.IP) (routeSpec *netlink.Rout
 }
 
 func (n *linuxNodeHandler) updateDirectRoutes(oldCIDRs, newCIDRs []*cidr.CIDR, oldIP, newIP net.IP, firstAddition, directRouteEnabled bool) error {
+	var acc error
```
Doesn't look like there can be multiple errors in this function in the current incarnation.

ti-mo commented on pkg/datapath/linux/node.go:
```diff
@@ -302,28 +321,46 @@ func (n *linuxNodeHandler) updateDirectRoutes(oldCIDRs, newCIDRs []*cidr.CIDR, o
 
-func (n *linuxNodeHandler) deleteDirectRoute(CIDR *cidr.CIDR, nodeIP net.IP) {
+func (n *linuxNodeHandler) deleteDirectRoute(CIDR *cidr.CIDR, nodeIP net.IP) error {
+	var acc error
```
This can be declared later.

ti-mo commented on pkg/datapath/linux/node.go:
```diff
@@ -335,14 +372,17 @@ func (n *linuxNodeHandler) deleteDirectRoute(CIDR *cidr.CIDR, nodeIP net.IP) {
 	if err != nil {
 		log.WithError(err).Error("Unable to list direct routes")
-		return
+		acc = errors.Join(acc, fmt.Errorf("list direct routes %s: %w", familyStr, err))
```
`return fmt.Errorf(...`

ti-mo commented on pkg/datapath/linux/node.go:
```diff
@@ -335,14 +372,17 @@ func (n *linuxNodeHandler) deleteDirectRoute(CIDR *cidr.CIDR, nodeIP net.IP) {
+		acc = errors.Join(acc, fmt.Errorf("list direct routes %s: %w", familyStr, err))
+		return acc
 	}
 
```
Declare `acc` here.

pippolo84 commented on pkg/datapath/linux/route/route_linux.go:
```diff
@@ -267,7 +267,11 @@ func Upsert(route Route) error {
 	if err != nil {
 		if nexthopRouteCreated {
-			deleteNexthopRoute(route, link, routerNet)
+			if err2 := deleteNexthopRoute(route, link, routerNet); err2 != nil {
+				// TODO: If this fails, we may want to add some retry logic.
+				log.WithError(err2).
+					Errorf("unable to clean up nexthop route following failure to replace route")
+			}
```
Nit: since we don't need to reference the outer `err`, we can simply use again the `err` name here and shadow the outer one.

tommyp1ckles replied:
err gets returned, so that would be changing the potential behaviour here.

pippolo84 replied:
Are you referring to the return at line 276? If so, that won't be affected, since the scope of `err2` is limited to the inner `if`. [Here](https://go.dev/play/p/EOEmPxuSQKD) a similar example where the "outer" `err` gets shadowed with an inner `err` whose scope is bounded to the inner `if`.

Anyway, I agree that shadowing might be trickier to read, so if you prefer to use `err2` instead that's completely fine :+1:

ti-mo commented on pkg/datapath/linux/node.go:
```diff
@@ -505,6 +551,14 @@ func upsertIPsecLog(err error, spec string, loc, rem *net.IPNet, spi uint8) {
 	if err != nil {
 		scopedLog.WithError(err).Error("IPsec enable failed")
+
+		for _, acc := range accs {
```
This is a little questionable, this function now takes two errs.

Since you're updating many call sites already, why not wrap `err` and return it? Join it into `acc`/`errs` at the call site.

tommyp1ckles replied:
yup, much cleaner, updated!

ti-mo commented on pkg/datapath/linux/node.go:
```diff
@@ -996,36 +1070,38 @@ func (n *linuxNodeHandler) enableIPsec(newNode *nodeTypes.Node) {
 
-	n.enableIPsecIPv4(newNode, zeroMark)
-	n.enableIPsecIPv6(newNode, zeroMark)
+	acc = errors.Join(acc, n.enableIPsecIPv4(newNode, zeroMark), n.enableIPsecIPv6(newNode, zeroMark))
```
`return errors.Join(...`

ti-mo commented on pkg/datapath/linux/node.go:
```diff
@@ -1035,24 +1111,27 @@ func (n *linuxNodeHandler) enableIPsecIPv4(newNode *nodeTypes.Node, zeroMark boo
 		remoteIP := newNode.GetCiliumInternalIP(false)
 		if remoteIP == nil {
-			return
+			return acc
```
This is a subtle change in behaviour. If `ipsec.IPsecDefaultDropPolicy` above returns an error, this statement will now cause the function to also return non-nil, which could affect its callers. Not sure if that's intended.

tommyp1ckles replied:
I think this is an easy case to reason about, prior to this PR enableIPsecIPv4/enableIPsecIPv6 did not return errors, so in this case I think we're safe from the problem of a change in behaviour of nil error causes some divergent behaviour down the line

pippolo84 commented on pkg/datapath/linux/node.go:
```diff
@@ -509,6 +509,14 @@ func upsertIPsecLog(err error, spec string, loc, rem *net.IPNet, spi uint8) {
 		scopedLog.WithError(err).Error("IPsec enable failed")
+
+		for _, acc := range accs {
+			*acc = errors.Join(*acc, fmt.Errorf("create endpoint %s (local_ip=%s, rem=%s, spi=%d): %w",
```
Just a heads up that `errors.Join` [uses a newline](https://pkg.go.dev/errors#Join) to format the wrapped errors.
If this should not be convenient for our logging, an alternative is to simply use fmt.Errorf(...) with a `%w` specifier for each error to be wrapped (and then define whatever custom separator in the format string).

tommyp1ckles replied:
Good point, the goal here was to consume as part of module health status, but in the meantime I have added extra logging and the node manager level so this could produce some ugly logs (I believe the newlines would be escaped by logrus though?). 

Wrapping doesn't really make sense in this context since the %w directive is supposed to add more context on top of an existing error, whereas the goal here is to provide a list of errors to indicate where reconciliation procedure calls have failed.

Ultimately, we do want to avoid this accumulation logic and just have well defined procedures that return when they encounter an error, thus reducing the potential number of states massively.  But I want to do more discussion/proposal work for that first.

pippolo84 replied:
> (I believe the newlines would be escaped by logrus though?).

TBH, I'm not sure.

> Wrapping doesn't really make sense in this context since the %w directive is supposed to add more context on top of an existing error, whereas the goal here is to provide a list of errors to indicate where reconciliation procedure calls have failed.

Actually, since Go 1.20 you can use `fmt.Errorf` with multiple `%w` directives, just to wrap multiple errors and customize the wrapped error string representation, like this:

`fmt.Errorf("%w-%w", err1, err2)`

That's basically equivalent of just wrapping `err1` and `err2` with `errors.Join`, but with the additional possibility of getting rid of the newline and specify a custom separator.

From the Go 1.20 release notes:

> The [fmt.Errorf](https://tip.golang.org/pkg/fmt/#Errorf) function now supports multiple occurrences of the %w format verb, which will cause it to return an error that wraps all of those error operands.
>
> The new function [errors.Join](https://tip.golang.org/pkg/errors/#Join) returns an error wrapping a list of errors.

Relevant code for [fmt.Errorf](https://cs.opensource.google/go/go/+/refs/tags/go1.20.5:src/fmt/errors.go) and for [errors.Join](https://cs.opensource.google/go/go/+/refs/tags/go1.20.5:src/fmt/errors.go).

> But I want to do more discussion/proposal work for that first.

Sounds good, then, let's go with this logic for now :+1:

ti-mo commented on pkg/datapath/linux/node.go:
```diff
@@ -1196,34 +1296,60 @@ func (n *linuxNodeHandler) nodeUpdate(oldNode, newNode *nodeTypes.Node, firstAdd
-		n.updateDirectRoutes(oldAllIP4AllocCidrs, newAllIP4AllocCidrs, oldIP4, newIP4, firstAddition, n.nodeConfig.EnableIPv4)
-		n.updateDirectRoutes(oldAllIP6AllocCidrs, newAllIP6AllocCidrs, oldIP6, newIP6, firstAddition, n.nodeConfig.EnableIPv6)
-		return nil
+		if err := func() error {
```
Instead of new acc and closure: `return errors.Join(n.updateDirectRoutes(), n.updateDirectRoutes())`...

tommyp1ckles replied:
The reasoning here, as well as the separate errors, is to prevent duplication when a series of function calls are all to be similarly "scoped".

For example

```go
errs = errors.Join(errs, fmt.Errorf("tried some thing: %w", foo1()))
//...
errs = errors.Join(errs, fmt.Errorf("tried some thing: %w", foo2()))
//...
errs = errors.Join(errs, fmt.Errorf("tried some thing: %w", foo3()))
// ...
```

Perhaps the closure here is a bit out of place, I didn't want to go and start refactoring sections of this into separate functions so I was trying to settle between the three approaches you see (somewhat inconstantly):

* create separate error accumulator.
* use a closure, wrap one error
* duplicate the same string around.

I'll try to make this more consistent for now,  I think we can go with "good enough" with approach three for now.

tommyp1ckles replied:
Cleaned this up to be more consistent with the rest of the changes.

ti-mo commented on pkg/datapath/linux/node.go:
```diff
@@ -1196,34 +1296,60 @@ func (n *linuxNodeHandler) nodeUpdate(oldNode, newNode *nodeTypes.Node, firstAdd
 	}
 
+	// For use in aggregating encapsulation code errors.
+	var encapAcc error
```
Is this really necessary? This just makes things less clear, and it's joined back into `acc` later.. I'd probably add extra wrapping to `updateTunnelMapping` calls instead.

tommyp1ckles replied:
See: Comment about the closure above for motivation, ill take another look to see if I can make things more consistent 😄

ti-mo commented on pkg/datapath/linux/node.go:
```diff
@@ -1301,33 +1440,45 @@ func (n *linuxNodeHandler) nodeDelete(oldNode *nodeTypes.Node) error {
+		}
 	}
 
+	var encapAcc error
```
Same as above, this is a little confusing.

pippolo84 commented on pkg/datapath/linux/node.go:
```diff
@@ -1196,34 +1296,60 @@ func (n *linuxNodeHandler) nodeUpdate(oldNode, newNode *nodeTypes.Node, firstAdd
+		}(); err != nil {
+			acc = errors.Join(acc, fmt.Errorf("enable direct routing: %w", err))
+		}
+		return acc
```
`errors.Join` is variadic, we can rewrite this in a (slightly) simpler way:

```suggestion
if err := func() error {
			return errors.Join(
				n.updateDirectRoutes(oldAllIP4AllocCidrs, newAllIP4AllocCidrs, oldIP4, newIP4, firstAddition, n.nodeConfig.EnableIPv4),
				n.updateDirectRoutes(oldAllIP6AllocCidrs, newAllIP6AllocCidrs, oldIP6, newIP6, firstAddition, n.nodeConfig.EnableIPv6),
			)
		}(); err != nil {
			return errors.Join(acc, fmt.Errorf("enable direct routing: %w", err))
		}
		return acc
```

tommyp1ckles replied:
I refactored this as part of timos request

ti-mo commented on pkg/datapath/linux/node.go:
```diff
@@ -1301,33 +1440,45 @@ func (n *linuxNodeHandler) nodeDelete(oldNode *nodeTypes.Node) error {
 	if n.nodeConfig.EnableIPSec {
-		n.deleteIPsec(oldNode)
+		if err := n.deleteIPsec(oldNode); err != nil {
+			acc = errors.Join(acc, fmt.Errorf("ipsec: %w", err))
```
```suggestion
			acc = errors.Join(acc, fmt.Errorf("delete ipsec: %w", err))
```

pippolo84 commented on pkg/datapath/linux/node.go:
```diff
@@ -1244,35 +1370,47 @@ func (n *linuxNodeHandler) nodeUpdate(oldNode, newNode *nodeTypes.Node, firstAdd
 
-		nodeID := n.allocateIDForNode(newNode)
+		nodeID, err := n.allocateIDForNode(newNode)
+		encapAcc = errors.Join(encapAcc, err)
```
It is possible to declare the encapsulation accumulator here directly:

```
nodeId, encapAcc := n.allocateIDForNode(newNode)
```

and get rid of the declaration at line 1352.

Also, I'd use something like `encapErrs` as a name for that (see other comment about "accumulator" naming)

ti-mo reviewed (dismissed):
Looks good overall, thanks! I think we should hold off on merging this until 1.15, this might introduce subtle changes in control flow despite all the care that was taken.

pippolo84 commented on pkg/datapath/linux/node.go:
```diff
@@ -1430,33 +1431,45 @@ func (n *linuxNodeHandler) nodeDelete(oldNode *nodeTypes.Node) error {
+		}
 	}
 
+	var encapAcc error
```
Same here

pippolo84 commented on pkg/datapath/linux/node.go:
```diff
@@ -158,7 +158,7 @@ func updateTunnelMapping(oldCIDR, newCIDR cmtypes.PrefixCluster, oldIP, newIP ne
 		if err := deleteTunnelMapping(oldCIDR, false); err != nil {
 			acc = errors.Join(acc,
-				fmt.Errorf("delete tunnel mapping (oldCIDR: %s): %w", oldCIDR, newIP))
+				fmt.Errorf("delete tunnel mapping (oldCIDR: %s, newIP: %s): %w", oldCIDR, newIP, err))
```
This fix here should be squashed in previous commit 98c694630497c0189eb47b6287b8acfaeb68d4f8.

pippolo84 commented on pkg/datapath/linux/node.go:
```diff
@@ -227,7 +227,7 @@ func createDirectRouteSpec(CIDR *cidr.CIDR, nodeIP net.IP) (routeSpec *netlink.R
 	routes, err = netlink.RouteGet(nodeIP)
 	if err != nil {
-		err = fmt.Errorf("unable to lookup route for node %s: %s", nodeIP, err)
+		err = fmt.Errorf("unable to lookup route for node %s: %w", nodeIP, err)
```
This change should be squashed in commit c076ddcfd45ca763f28914caab68c7596f34bc77

pippolo84 commented on pkg/node/manager/manager_test.go:
```diff
@@ -833,3 +840,51 @@ func (s *managerTestSuite) TestNode(c *check.C) {
+
+	// Start the manager
+	assert.NoError(m.Start(context.Background()))
+	time.Sleep(time.Millisecond * 100)
```
Why we need this `time.Sleep` here? I am afraid that this might increase test flakyness when running in CI.

tommyp1ckles replied:
oops, leftover from old iteration, removed!

pippolo84 reviewed (changes requested):
Thanks for the PR! I've left some suggestions inline, mainly around naming and style.

christarazi commented on pkg/node/manager/manager.go:
```diff
@@ -292,13 +308,25 @@ func (m *manager) backgroundSync(ctx context.Context) error {
 
+		if errs != nil {
+			log.WithError(errs).Debug("Failed to do node validation")
+			m.healthReporter.Degraded("Failed to do node validation", errs)
```
Nit: `Failed to perform node validation`

christarazi replied:
Do we need both debug logs and health reporter logs? Maybe we can put a debug log inside the health reporter instead of calling `.Debug` at each call site of the health reporter.

tommyp1ckles replied:
Probably not actually, the health reporter has debug logging on emitting status updates (which we may consider making into info at some point).

christarazi commented on pkg/node/manager/manager.go:
```diff
@@ -292,13 +308,25 @@ func (m *manager) backgroundSync(ctx context.Context) error {
 			m.Iter(func(nh datapath.NodeHandler) {
-				nh.NodeValidateImplementation(entry.node)
+				if err := nh.NodeValidateImplementation(entry.node); err != nil {
+					log.WithError(err).Errorf("Failed to validate handler %s on node %s", nh.Name(), entry.node.Name)
```
Let's use structured logging (`WithFields`) and avoid dynamic log formatting. See `pkg/logging/logfields` as well for existing logfields and add any that are needed.

christarazi commented on pkg/node/manager/manager.go:
```diff
@@ -122,6 +126,10 @@ type manager struct {
+
+	healthReporter cell.HealthReporter
+
+	syncObs stream.Observable[any]
```
Let's stick to the convention in this struct to have godocs on each field. Please also elaborate on `syncObs`.

christarazi commented on pkg/node/manager/manager.go:
```diff
@@ -133,7 +141,9 @@ func (m *manager) Subscribe(nh datapath.NodeHandler) {
 		v.mutex.Lock()
-		nh.NodeAdd(v.node)
+		if err := nh.NodeAdd(v.node); err != nil {
+			log.WithError(err).Errorf("handler initial subscribe: %s failed on node %s", nh.Name(), v.node.Name)
```
For error and warning level logs, we need to be more thoughtful about the log msgs. The msgs need to be useful for the end user by telling them what are the ramifications and what Cilium will do about it. In some cases there's nothing that a user can do and so for example, the msg can say to file an issue.

In this case, I'd suggest 

```
Initial node event handling failed. Cilium may be in a degraded state. See error message for more details.
```

In the future, we can improve the error msg to say that Cilium will retry, etc. 

Please apply the above to the rest of the PR for errors / warnings logs.

christarazi commented on pkg/datapath/linux/node.go:
```diff
@@ -509,9 +509,15 @@ func upsertIPsecLog(err error, spec string, loc, rem *net.IPNet, spi uint8) {
+			loc.String(),
+			rem.String(),
+			spi, err)
 	} else {
```
I think we should avoid the custom formatting (`local_ip=...`) and instead make the error msg itself more naturally flow:

```
return fmt.Errorf("failed to create endpoint for %q with local IP %q and remote IP %q: %w", spec, loc, rem)
```

christarazi replied:
Not sure what "create endpoint" is referring to here though

christarazi replied:
Also, if we are already going to log when this error occurs, then we don't need to stuff the error msg with the same context again. It would also be good to use the newly created error within the log msg, i.e. `WithError()`

tommyp1ckles replied:
I was in between two priorities in many of these:

* keeping the error messages as terse/short as possible as to avoid excessively long errors once they've made their way up the stack.
* Providing enough context info such that users aren't left scratching their heads about why exactly the error occured.

christarazi commented on pkg/datapath/linux/node.go:
```diff
@@ -319,6 +319,15 @@ func (n *linuxNodeHandler) updateDirectRoutes(oldCIDRs, newCIDRs []*cidr.CIDR, o
 	for _, cidr := range addedCIDRs {
 		if routeSpec, err := installDirectRoute(cidr, newIP); err != nil {
 			log.WithError(err).Warningf("Unable to install direct node route %s", routeSpec.String())
+			// In the current implementation, this often fails because updates are tried for both ip families
+			// regardless if the Node has either ip types.
+			// At the time of this change we are only interested in bubbling up errors without affecting execution flow.
+			// Thus we are ignoring the error here for now.
+			//
+			// TODO(Tom): In the future we will want to avoid attempting to do the update if we know it will fail.
+			if newIP == nil && errors.Is(err, unix.ERANGE) {
+				return nil
+			}
```
My concern with this is that we end up ignoring legitimate errors, but maybe this stems from the fact that I don't understand fully why this commit is needed.

tommyp1ckles replied:
Yeah, it's possible, but right now this error happens every time in some situations.  There's a better solution to this but it involves avoiding trying certain calls if ip4/ip6 isn't enabled.

christarazi commented on pkg/datapath/linux/node.go:
```diff
@@ -1473,102 +1622,117 @@ func (n *linuxNodeHandler) createNodeExternalIPSecOutRoute(ip *net.IPNet, dflt b
 
 	if err := route.Upsert(n.createNodeExternalIPSecOutRoute(ip, true)); err != nil {
 		log.WithError(err).WithField(logfields.CIDR, ip).Error("Unable to replace the IPSec route OUT the default routing table")
+		errs = errors.Join(errs, fmt.Errorf("replace ipsec default route OUT: %w", err))
```
One nit about the new error msgs: they should state exactly what went wrong so that they can be joined together in the natural way that Go errors are aggregated.

Here, it should say:

```suggestion
		errs = errors.Join(errs, fmt.Errorf("failed to replace ipsec default route in the OUT direction: %w", err))
```

In theory, error msgs should be read like:

```
failed to validate Linux node: failed to process new Linux node event: failed to replace ipsec default route in the OUT direction: ENOMEM
```

Please ensure all new error msgs adhere to the above.

tommyp1ckles replied:
When I was working on this I was concerned about creating unnecessary redundancy in errors, for example the "failed to..." gets repeated a lot down the stack and doesn't add much information.

I would argue that an error implies that something has failed, IMO to make things more clear we should add this wording closer to the top level:

```
failed to validate Linux node: processing new event: replacing ipsec default route in the OUT direction: ENOMEM
```

Where the top level explicitly indicates that there has been a failure/error/problem etc.  Following this you have essentially a sort of stack trace indicating error context all the way down.  With finally the error terminating with the core error.

(I also was trying to avoid having words like "node" or "route" repeated all the way down the stack).

christarazi replied:
I think my main principle here is to be consistent with the rest of the Cilium code base in regards to how errors are formatted. We need to think about whether what the PR code does is also how each module should structure / format their errors. Otherwise, if we follow how the rest of the code base is, then there's no "controversy" so to speak. WDYT?

tommyp1ckles replied:
makes sense, ill keep things consistent with the broader code base here for now, we can continue to figure out the best way to do these messages going forward.

christarazi reviewed (commented):
Nice work!

I have a smattering of random comments related to formatting the errors and logs so that we aren't introducing new signals into the tree which may be incorrectly interpreted by us as developers or users.

In regards to the discussion about control flow, I agree that we should avoid as many side effects as possible. My inclination is that this change can be done in a way that does not contain any additional side effects. If that's the case, I'd like to understand more deeply why that's the case. However, as long as we avoid changing the control flow and this PR effectively becomes just a change that better reports the status / errors of the node manager, then it can still be considered for v1.14.

pippolo84 commented on pkg/datapath/linux/node.go:
```diff
@@ -1165,6 +1253,10 @@ func (n *linuxNodeHandler) subnetEncryption() bool {
 	var (
+		// Don't stop executing the function if we get an error. Instead we
+		// log and aggregate errors in accumulator.
+		acc error
```
I'd change this one too to `errs`

pippolo84 commented on pkg/node/manager/manager_test.go:
```diff
@@ -829,3 +840,50 @@ func (s *managerTestSuite) TestNode(c *check.C) {
 	c.Assert(n, checker.DeepEquals, *n1V2)
 }
+
+func TestNodeManagerEmitStatus(t *testing.T) {
```
Trying the test locally, it seems to be flaky:

```
$ go test -count=5 -v ./pkg/node/manager/... -run="TestNodeManagerEmitStatus"
=== RUN   TestNodeManagerEmitStatus
--- FAIL: TestNodeManagerEmitStatus (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x1fe4738]

goroutine 22 [running]:
...
github.com/cilium/cilium/pkg/node/manager.TestNodeManagerEmitStatus(0x0?)
	/home/pippolo/go/src/github.com/cilium/cilium/pkg/node/manager/manager_test.go:869 +0x618
...
FAIL	github.com/cilium/cilium/pkg/node/manager	0.692s
FAIL
```

The panic happens when `m.syncObs` has not yet been initialized (that's because `backgroundSync` is called asynchronously from the workerpool). A solution might be to initialize syncObs in `*manager.Start`, save a reference to the returned `tick` and `complete` functions inside the manager struct to be called later where needed.

Also, I think the test can be greatly simplified avoiding the creation of multiple channels and doing this:

```
	ch := make(chan struct{})
	m.syncObs.Observe(ctx, func(_ any) {
		ch <- struct{}{}
	}, func(err error) {
		close(ch)
	})
	tick := func() {
		<-ch
	}
```

As a final consideration, I don't like too much the idea of adding code into the manager that has the sole purpose of testing the integration with the health reporter (unless we are planning to use `syncObs` for something else in the future).
What about removing the `stream.Multicast` and write the test like this:

```
	// Start the manager
	assert.NoError(m.Start(ctx))

	require.Eventually(t, func() bool { return hr.Err() != nil }, time.Minute, 5*time.Millisecond)
	nh1.NodeValidateImplementationEventError = nil
	require.Eventually(t, func() bool { return hr.Err() == nil }, time.Minute, 5*time.Millisecond)

	// Stop the manager
	assert.NoError(m.Stop(ctx))
```

tommyp1ckles replied:
Hmm yeah I didn't know about require.Eventually, I was hesitant to add the stream but I wanted to avoid non determinism by just using `time.Sleep()`

I'm going to change this over to just use the require approach, should clean things up quite a bit

tommyp1ckles replied:
Oof this one is tricker than expected, upon further inspection, even this is a bit racey and lead to flakes.

I ended up just using the existing test handler fixture, but syncing on handler updates, this method proved a lot more reliable since we can sync on both the handler update and the health status update.

I ran it for a million iterations without a flake so hopefully this is the best way forward.
```
go test -failfast -count=1000000 -v -run=TestNodeManagerEmitStatus
```

pippolo84 reviewed (changes requested):
Thanks for addressing the previous comments. I've left a comment about the manager test.

pippolo84 reviewed (approved):
Awesome, thanks! :rocket:

brb reviewed (approved):
WireGuard changes :ok_hand:

pippolo84 reviewed (approved):
Recent changes LGTM :rocket:

brb commented on pkg/datapath/linux/node.go:
```diff
@@ -509,9 +509,15 @@ func upsertIPsecLog(err error, spec string, loc, rem *net.IPNet, spi uint8) {
 	})
 	if err != nil {
 		scopedLog.WithError(err).Error("IPsec enable failed")
+		return fmt.Errorf("failed to enable ipsec with %s using local IP %s, rem %s, spi %d: %w",
```
I think the error logging above (and in other places) is no longer needed, as it will be logged by the node manager.

brb commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -536,42 +537,48 @@ func matchesOnNodeID(mark *netlink.XfrmMark) bool {
 	if err != nil {
 		scopedLog.WithError(err).Warning("Failed to list XFRM states for deletion")
-		return
+		return err
```
Ditto with the above.

brb reviewed (approved):
Nice changes, thanks! I guess we can remove the redundant logging in a follow-up PR.
